node/src/Pos/Wallet/Web/Pending/Functions.hs

```diff
@@ -0,0 +1,181 @@
+reevaluateUncertainPtxs ptxs =  withStateLock LowPriority "reevaluateUncertainPtxs" $ \__tip -> do
+    -- Get all transactions we're uncertain about
+    let isUncertain ptx = has (ptxCond . _PtxApplying)  ptx ||
+                          has (ptxCond . _PtxWontApply) ptx
+    let (uncertain, certain) = partitionHashMap isUncertain ptxs
+    -- Find those that are already in blocks; they'll be marked as
+    -- 'PtxInNewestBlocks'
+    presentIds <- getTxsBlockStatus @ssc (getKeys uncertain)
+    let present :: HashMap TxId (PendingTx, ChainDifficulty)
+        present = HM.intersectionWith (,) uncertain presentIds
+    let missing :: HashMap TxId PendingTx
+        missing = HM.difference uncertain presentIds
+    -- Topsort the rest ('missing'), go one by one and either add to UTXO or
+    -- mark as 'PtxWontApply'
+    --
+    -- TODO: is evalToilTEmpty okay?
+    (missingInvalid, missingValid) <- evalToilTEmpty $ do
+        let sorted = fromMaybe (error "Pending transacrions couldn't be topsorted") $
+                     topsortTxs ptxWithHash (toList missing)
+        fmap partitionEithers $ forM sorted $ \ptx -> do
+            let vtc = VTxContext True     -- TODO: is it necessarily 'True'?
+            runExceptT (verifyTxUtxo vtc (ptx ^. ptxTxAux)) >>= \case
+                Left err -> pure (Left (ptx, err))
+                Right _  -> applyTxToUtxo (ptxWithHash ptx) >>
+                            pure (Right ptx)
+
+    logInfo $ sformat ("These transactions are certain and won't be affected: "
+                       %listJson) (HM.keys certain)
+    logInfo $ sformat ("These transactions are present in blocks: "
+                       %listJson) (HM.keys present)
+    logInfo $ sformat ("These transactions are not in blocks and invalid: "
+                       %listJson) (map (view ptxTxId . fst) missingInvalid)
+    logInfo $ sformat ("These transactions are not in blocks and valid: "
+                       %listJson) (map (view ptxTxId) missingValid)
+
+
+    pure $ mconcat
+        [ certain                          -- certain = do nothing
+        , map setInNewestBlocks present    -- present = PtxInNewestBlocks
+        , mkHashMap (view ptxTxId) $       -- invalid = PtxWontApply
+            map setWontApply missingInvalid
+        , mkHashMap (view ptxTxId) $       -- valid = PtxApplying
+            map setApplying missingValid
+        ]
+
+----------------------------------------------------------------------------
+-- General utilities
+----------------------------------------------------------------------------
+
+-- | Create a 'HashMap' by deriving a key from each value.
+mkHashMap
+    :: (Eq k, Hashable k)
+    => (v -> k) -> [v] -> HashMap k v
+mkHashMap f = HM.fromList . map (f &&& identity)
+
+-- | Partition a 'HashMap'.
+partitionHashMap
+    :: (v -> Bool) -> HashMap k v -> (HashMap k v, HashMap k v)
+partitionHashMap p hm = (HM.filter p hm, HM.filter (not . p) hm)
```

Comment:
Wouldn't be quicker to filter the first and then write the second as a `HM.difference` on the input map?

Reply:
Possibly, I haven't benchmarked it.

NB. I've avoided “decorate-sort-undecorate” on purpose because `p` is a cheap function in all cases where `partitionHashMap` is used.